### PR TITLE
[WIP][SPARK-25305][SQL] Respect attribute name in CollapseProject and ColumnPruning

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Currently in optimizer rule `CollapseProject`, the lower level project is collapsed into upper level, but the naming of alias in lower level is propagated in upper level.
In `ColumnPruning`,  `Project` is eliminated if its child's output attributes is `semanticEquals` to it, even the naming doesn't match.

Let's see the follow example:
```
        val location = "/tmp/t"
        val df = spark.range(10).toDF("id")
        df.write.format("parquet").saveAsTable("tbl")
        spark.sql("CREATE VIEW view1 AS SELECT id FROM tbl")
        spark.sql(s"CREATE TABLE tbl2(ID long) USING parquet location $location")
        spark.sql("INSERT OVERWRITE TABLE tbl2 SELECT ID FROM view1")
        println(spark.read.parquet(location).schema)
        spark.table("tbl2").show()
```
The output column name in schema will be `id` instead of `ID`, thus the last query shows nothing from `tbl2`. 
By enabling the debug message we can see that the output naming is changed from `ID` to `id`, and then the `outputColumns` in `InsertIntoHadoopFsRelationCommand` is changed in `RemoveRedundantAliases`.
wechatimg5

wechatimg4

With the fix proposed in this PR, the output naming `ID` won't be changed.
wechatimg3

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
Thanks @wangyum for finding the issue. Before this PR, he propose https://github.com/apache/spark/pull/22124 and https://github.com/apache/spark/pull/22287 . 
But there is no description for the key reason of regression provided, also the test case is complex.
So I created this one.
We may also need to change the `outputColumns` in `DataWritingCommand` afterward, to avoid it is changed after optimization.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -649,17 +648,20 @@ object CollapseProject extends Rule[LogicalPlan] {
...
 
-  private def collectAliases(projectList: Seq[NamedExpression]): AttributeMap[Alias] = {
-    AttributeMap(projectList.collect {
-      case a: Alias => a.toAttribute -> a
+  private def collectAliases(
+      upper: Seq[NamedExpression], lower: Seq[NamedExpression]): AttributeMap[Alias] = {
+    AttributeMap(lower.zipWithIndex.collect {
+      case (a: Alias, index: Int) =>
+        a.toAttribute ->
+          a.copy(name = upper(index).name)(a.exprId, a.qualifier, a.explicitMetadata)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Do we have a better way for Alias.copy(...) ?

**Comment by gengliangwang** [comment]
@cloud-fan @gatorsmile @wangyum

**Comment by maropu** [comment]
This behaivour depends on `spark.sql.caseSensitive`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -515,8 +515,7 @@ object PushProjectionThroughUnion extends Rule[LogicalPlan] with PredicateHelper
  */
 object ColumnPruning extends Rule[LogicalPlan] {
   private def sameOutput(output1: Seq[Attribute], output2: Seq[Attribute]): Boolean =
-    output1.size == output2.size &&
-      output1.zip(output2).forall(pair => pair._1.semanticEquals(pair._2))
+    output1.size == output2.size && output1.zip(output2).forall(pair => pair._1 == pair._2)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think still we need to check if the exprIds that they refere to are the same.

**Comment by cloud-fan** [comment]
> This behaivour depends on spark.sql.caseSensitive?

No. It's writing not resolving a column, so Spark should be case-preserving.

**Comment by maropu** [comment]
ok, thanks.

**Comment by gengliangwang** [comment]
> We may also need to change the outputColumns in DataWritingCommand afterward, to avoid it is changed after optimization.
After second thought, I will create PR to fix `outputColumns` first. It takes time to cover all the cases by changing optimizer rules.
Mark this PR as "WIP" for now.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by maropu** [comment]
Can this pr include the revert code for the #22320 change? Or, we should keep the change for safeguard for a while? (I thinks we'd be better to choose the second approach, so this is just a question).

**Comment by gengliangwang** [comment]
@maropu Let's keep the change of #22320 for a while. Thanks.

**Comment by maropu** [comment]
ok, thanks for the check.

**Pull request closed without merging**